### PR TITLE
feat: [2/2] add protoc executable for platforms

### PR DESCRIPTION
**Pull request by xiaozhenliu-gg5**
1. Add protoc binary in `test/tools` so that we don't have to depend on the pre-installed protoc command line library. Windows/Linux/OS equivalents are included.
2. Fix unit tests messages.

This should be hold until #3 merged.

### CODE REVIEW

File: .flaky8 [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  # Closing bracket mismatches opening bracket's line.
+  # This works poorly with type annotations in method declarations.
+  E123, E124
+  # Line over-indented for visual indent.
+  # This works poorly with type annotations in method declarations.
+  E126, E128, E131
+  # Line break after binary operator.
+  # This catches line breaks after "and" / "or" as a means of breaking up
+  # long if statements, which PEP 8 explicitly encourages.
+  W504
```
Comment (software-dov) [inline .flaky8:12]: I'd like to add my vote for just using [black](https://github.com/psf/black) to enforce styling and formatting, but that is outside the scope of this review.

File: .github/workflows [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  pull_request:
+    branches: [ $default-branch ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.5, 3.6, 3.7, 3.8]
```
Comment (software-dov) [inline .github/workflows:18]: 3.9 was also released just this past week. Should we add it too?

File: src/comparator/enum_comparator.py [hunk 3]
```diff
@@ -3,42 +3,62 @@
...
-            enum_values_dict_update = {x.number: x for x in self.enum_update.values}
+            enum_values_dict_original = {
+                x.number: x for x in self.enum_original.value}
+            enum_values_dict_update = {
+                x.number: x for x in self.enum_update.value}
             # Compare Enum values that only exist in original version
-            for number in list(set(enum_values_dict_original.keys()) - set(enum_values_dict_update.keys())):
-                EnumValueComparator(enum_values_dict_original[number], None).compare()
+            for number in list(set(enum_values_dict_original.keys()) -
+                               set(enum_values_dict_update.keys())):
```
Comment (software-dov) [inline src/comparator/enum_comparator.py:50]: Nit: outside the scope of this review (which mostly seems to be formatting), but we can iterate over sets directly without having to turn them into lists (and we also don't need to call `keys`; keys is the implicit iteration over a dictionary):
```python
for n in set(enum_values_dict_original) - set(enum_values_dict_update):
```

I'm not going to tag all the instances, but the nit applies to all `list(set())` constructions.

File: src/comparator/field_comparator.py [hunk 4]
```diff
@@ -1,63 +1,68 @@
...
+                self.field_update.type)
+            msg = 'Type of the field is changed, the original is {}, but the updated is {}'.format(
+                type_original, type_update)
+            FindingContainer.addFinding(
+                FindingCategory.FIELD_TYPE_CHANGE, "", msg, True)
 
+        # 6. Check `google.api.resource_reference` annotation.
+        # TODO(xiaozhenliu): annotation is removed, but the using
+        # file-level resource is added to the message.
+        # This should not be taken as breaking change.
```
Comment (software-dov) [inline src/comparator/field_comparator.py:68]: Really? Is there documentation for that somewhere?

File: src/comparator/message_comparator.py [hunk 5]
```diff
@@ -3,62 +3,91 @@
...
+            self,
+            nested_msg_dict_original,
+            nested_msg_dict_update
+    ):
+        msg_unique_original = list(
+            set(nested_msg_dict_original.keys()) - set(nested_msg_dict_update.keys()))
+        msg_unique_update = list(
+            set(nested_msg_dict_update.keys()) - set(nested_msg_dict_original.keys()))
+        msgIntersaction = list(
+            set(nested_msg_dict_original.keys()) & set(nested_msg_dict_update.keys()))
```
Comment (software-dov) [inline src/comparator/message_comparator.py:84]: Nit: since we're using the same two sets three times each, why not lift them into variables?
```python
original = set(nested_msg_dict_original)
update = set(nested_msg_dict_update)
msg_unique_original = original - update
msg_unique_update = update - original
msg_intersection = original & update
```

File: src/comparator/service_comparator.py [hunk 6]
```diff
@@ -31,61 +33,86 @@ def compare(self):
...
+        method: MethodDescriptorProto
+    ) -> bool:
         # (AIP 158) The response must not be a streaming response.
         if method.server_streaming:
             return False
         responseMsg = method.output_type
         # API must provide a `string next_page_token` field.
-        if ('next_page_token' not in responseMsg.fields_by_name) or responseMsg.fields_by_name['next_page_token'].type != 9:
+        if ('next_page_token' not in responseMsg.fields_by_name) or \
+                responseMsg.fields_by_name['next_page_token'].type != 9:
```
Comment (software-dov) [inline src/comparator/service_comparator.py:110]: Very minor nit: the `\` isn't necessary if you wrap the entire conditional in parentheses.

File: test/tools/invoker.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            subprocess.run(['rm', descriptor_set_output])
+
+    @classmethod
+    def _get_protoc_binary(cls) -> str:
+        if platform.system == 'Windows':
+            return os.path.join(cls._CURRENT_DIR, 'test/tools/protoc.exe')
+        elif platform.system == 'Linux':
+            return os.path.join(cls._CURRENT_DIR, 'test/tools/protoc')
+        else:
+            return os.path.join(cls._CURRENT_DIR, 'test/tools/osx-protoc')
```
Comment (software-dov) [inline test/tools/invoker.py:60]: This doesn't handle all the cases (e.g. Java); would it be better to do something like
```python
systems_to_name = {
    "Windows": "protoc.exe",
    "Linux": "protoc",
    "Darwin": "osx-protoc",
}
name = systems_to_name.get(platform.system)
if not name:
    raise ...

return os.path.join(cls._CURRENT_DIR, "test", "tools", name)
```
?

File: test/tools/invoker.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class _ProtocInvokerException(Exception):
+    def __init__(self, message):
+
+        # Call the base class constructor with the parameters it needs
+        super().__init__(message)
```
Comment (software-dov) [inline test/tools/invoker.py:67]: See my comment in the other review about just making the body `pass` if we're not really changing the constructor.

**Review by software-dov — COMMENTED** [review]
Looks great. It's really coming together!

File: src/comparator/enum_comparator.py (hunk 3 above)
Comment (xiaozhenliu-gg5) [reply src/comparator/enum_comparator.py:50]: You are right! Added this change in this #2 PR. Thanks!

File: src/comparator/message_comparator.py (hunk 5 above)
Comment (xiaozhenliu-gg5) [reply src/comparator/message_comparator.py:84]: Sounds reasonable, also modified this in #2

File: test/tools/invoker.py (hunk 8 above)
Comment (xiaozhenliu-gg5) [reply test/tools/invoker.py:67]: 👍 non-format changes are included in this #2 and will rebase this PR to get the new changes once #2 is merged. Thank you!

File: .flaky8 (hunk 1 above)
Comment (xiaozhenliu-gg5) [reply .flaky8:12]: Thanks! Will bring the discussion to #3 and update from there.

File: .github/workflows (hunk 2 above)
Comment (xiaozhenliu-gg5) [reply .github/workflows:18]: That's great! Thanks for the news, added the newest version in #3

File: src/comparator/service_comparator.py (hunk 6 above)
Comment (xiaozhenliu-gg5) [reply src/comparator/service_comparator.py]: Noted, this is helpful! Added parentheses to remove the `/`.

File: src/comparator/field_comparator.py (hunk 4 above)
Comment (xiaozhenliu-gg5) [reply src/comparator/field_comparator.py:69]: Sorry for the confusion, for my understanding from the design [doc](https://docs.google.com/document/d/16MrkP4C2Rp_wnFuZjV6uubm5Kt61BYcPQwfyFMugOD8/edit#) If the `google.api.resource_reference` annotation is removed, it is a breaking change unless the reference was to itself, using a file-level resource annotation, and the file-level annotation has been moved into this message.

```
option (google.api.resource_definition) = {
  type: "iam.googleapis.com/ServiceAccount"
  pattern: "projects/{project}/serviceAccounts/{service_account}"
};

message GenerateAccessTokenRequest {
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference).type = "iam.googleapis.com/ServiceAccount",
  ];
  .....
}
// .... nowhere else uses ServiceAccount resource definition.
```
This should be equivalent to 
```
// file-level resource definition is removed.
message GenerateAccessTokenRequest {
  option (google.api.resource) = {
    type: "iam.googleapis.com/ServiceAccount",
    pattern: "projects/{project}/serviceAccounts/{service_account}"
  };
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference).type = "iam.googleapis.com/ServiceAccount",
  ];
  .....
}
```

**Comment by xiaozhenliu-gg5** [comment]
We decide not to support `Windows` platform ATM, so removed the`protoc.exe` in the test tools.

File: src/comparator/field_comparator.py (hunk 4 above)
Comment (chrisdunelm) [reply src/comparator/field_comparator.py]: Yes, I believe this is correct, and is as specified in the doc.

File: test/tools/invoker.py [hunk 9]
```diff
@@ -50,6 +51,18 @@ def cleanup(self):
         if os.path.exists(descriptor_set_output):
             os.remove(descriptor_set_output)
 
+    @classmethod
+    def _get_protoc_binary(cls) -> str:
+        system_to_protoc_binary = {
+            "Linux": "protoc",
+            "Darwin": "osx-protoc",
```
Comment (chrisdunelm) [inline test/tools/invoker.py:58]: Is this because you're developing it on a mac?

**Review by chrisdunelm — APPROVED** [review]
LGTM

File: test/tools/invoker.py (hunk 9 above)
Comment (xiaozhenliu-gg5) [reply test/tools/invoker.py:58]: Yes! But I can remove it and develop on cloudtop using Linux :)

Comment (xiaozhenliu-gg5) [reply test/tools/invoker.py:58]: Done. Thanks!
